### PR TITLE
nixos/eternal-terminal: init new module.

### PR DESCRIPTION
**Pull request by qolii**
###### Motivation for this change
`eternal-terminal` is like `ssh`; it requires a daemon to be running if you want to allow remote machines to connect to you. This module provides this.

I have based the generated systemd unit and config file on the files provided in the et repo ([systemd unit](https://github.com/MisterTea/EternalTerminal/blob/master/systemctl/et.service) and [cfg file](https://github.com/MisterTea/EternalTerminal/blob/master/etc/et.cfg)). However, if anyone has experience or expertise with better ways to configure it, I am all ears. In particular, I have made `verbose`, `silent` and `logsize` as module parameters, but these aren't really described on the ET website and I'm not sure exactly how to control them.

In addition, I think that when the remote `et` connects to the server, it logs in and then runs `etterminal`. Because of this, the `etterminal` binary needs to be available to it at that time. To solve this, I added the `eternal-terminal` package to `environment.systemPackages`, but this doesn't seem that elegant. If there is a better way, we can change to that.

This is the first unit I have made. So all feedback is very welcome.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Review by Ma27 — CHANGES_REQUESTED** [review]
first of all, thanks a lot for contributing a module for eternal terminal! The idea behind this tool seems quite interesting, I'm currently wondering if I could benefit from this as well :)

However I added some comments about the option's design in the module I'd like to see fixed before merging this.

File: nixos/modules/services/networking/eternal-terminal.nix [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+      enable = mkOption {
+        default = false;
+        type = types.bool;
+        description = ''
+          Enable the Eternal Terminal server.
+        '';
+      };
+
+      port = mkOption {
+        default = null;
```
[the remark this answers is not in the document]
Comment (qolii) [reply nixos/modules/services/networking/eternal-terminal.nix]: I was debating this with myself a little already. I am happy to change it though, I think you're right.

File: nixos/modules/services/networking/eternal-terminal.nix [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+        type = types.nullOr types.int;
+        description = ''
+          The port the server should listen on. Will use the server's default (2022) if not specified.
+        '';
+      };
+
+      verbosity = mkOption {
+        default = 0;
+        type = types.int;
```
[the remark this answers is not in the document]
Comment (qolii) [reply nixos/modules/services/networking/eternal-terminal.nix]: +1!

File: nixos/modules/services/networking/eternal-terminal.nix [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+        description = ''
+          The verbosity level (0-9).
+        '';
+      };
+
+      silence = mkOption {
+        default = 0;
+        type = types.int;
+        description = ''
+          Silence.
```
[the remark this answers is not in the document]
Comment (qolii) [reply nixos/modules/services/networking/eternal-terminal.nix]: Ah, I actually don't know what this one does, and I can't find it in the doc. It's simply used in the provided upstream config file :/

Perhaps another ET user might know? Definitely happy to upgrade this description.

**Comment by qolii** [comment]
@Ma27, thanks for your feedback! My system is suffering through a big rebuild, so I'll need to make these changes in a little bit. Back soon!

File: nixos/modules/services/networking/eternal-terminal.nix (hunk 1 above)
[the remark this answers is not in the document]
Comment (Ma27) [reply nixos/modules/services/networking/eternal-terminal.nix]: I certainly understand your point, I had to think about it either. But in the end I guess that this is the better for documentation purposes :)

File: nixos/modules/services/networking/eternal-terminal.nix (hunk 3 above)
[the remark this answers is not in the document]
Comment (Ma27) [reply nixos/modules/services/networking/eternal-terminal.nix]: I see.
Admittedly I didn't find anything useful in the docs as well last night. I'll have a look at the upstream sources, when we add this option, this needs some more documentation IMHO.

[the remark this answers is not in the document]
Comment (Ma27) [reply nixos/modules/services/networking/eternal-terminal.nix]: so, `silent` seems to be an option to disable all logs:
https://github.com/MisterTea/EternalTerminal/commit/53e409fefd72ce3d84c5c38ad2897c0f66d57ab5

File: nixos/modules/services/networking/eternal-terminal.nix [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+
+      verbosity = mkOption {
+        default = 0;
+        type = types.int;
+        description = ''
+          The verbosity level (0-9).
+        '';
+      };
+
+      silence = mkOption {
```
Comment (Ma27) [inline nixos/modules/services/networking/eternal-terminal.nix]: `silent` may be better here as this is the actual configuration parameter which is more consistent IMHO.

File: nixos/modules/services/networking/eternal-terminal.nix [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+        default = 0;
```
Comment (Ma27) [inline nixos/modules/services/networking/eternal-terminal.nix]: as this is a true/false toggle this should be of type boolean and then transformed with `if cfg.silent then "1" else "0"`.

Comment (qolii) [reply nixos/modules/services/networking/eternal-terminal.nix]: Yep.

File: nixos/modules/services/networking/eternal-terminal.nix (hunk 4 above)
Comment (qolii) [reply nixos/modules/services/networking/eternal-terminal.nix]: Agreed.

File: nixos/modules/services/networking/eternal-terminal.nix (hunk 3 above)
[the remark this answers is not in the document]
Comment (qolii) [reply nixos/modules/services/networking/eternal-terminal.nix]: Thanks for finding this. I'll add some extra words.

The argument situation with et looks like a real mess, documentation wise. It seems like there are loads of them. I think we could probably add more later, but for now this is a reasonable subset.

File: nixos/modules/services/networking/eternal-terminal.nix (hunk 1 above)
[the remark this answers is not in the document]
Comment (qolii) [reply nixos/modules/services/networking/eternal-terminal.nix]: :) Agreed.

File: nixos/modules/services/networking/eternal-terminal.nix [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+
+{
+
+  ###### interface
+
+  options = {
+
+    services.eternal-terminal = {
+
+      enable = mkOption {
```
[the remark this answers is not in the document]
Comment (qolii) [reply nixos/modules/services/networking/eternal-terminal.nix]: Will change.

**Comment by qolii** [comment]
@Ma27, I think I addressed all your comments so far. Let me know if there is anything else.

**Comment by Ma27** [comment]
the change seems fine so far. Please give me some time to test this on one of my servers to confirm its functionality (and I like the idea of this,so I'd like to try it out first:) ).

File: nixos/modules/services/networking/eternal-terminal.nix [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+          ExecStart = "${pkgs.eternal-terminal}/bin/etserver --daemon --cfgfile=${pkgs.writeText "et.cfg" ''
+            ; et.cfg : Config file for Eternal Terminal
+            ;
+
+            [Networking]
+            port = ${toString cfg.port}
+
+            [Debug]
+            verbose = ${toString cfg.verbosity}
+            silent = ${if cfg.silent then "true" else "false"}
```
Comment (Ma27) [inline nixos/modules/services/networking/eternal-terminal.nix]: shouldn't the values be 0/1?

**Comment by Ma27** [comment]
checked it on one of my servers with a recent 18.09 and the eternal-terminal changes (bump and module) cherry-picked. Except the issue above the module seems fine :)

File: nixos/modules/services/networking/eternal-terminal.nix (hunk 7 above)
Comment (qolii) [reply nixos/modules/services/networking/eternal-terminal.nix]: Changed!

**Comment by qolii** [comment]
@Ma27, thanks for your careful review! Do you think we are good to go? This is working well for me.

**Review by Ma27 — APPROVED** [review]
as mentioned earlier I've checked this on oneof my servers and it worked fine, code looks good as well, so this should be safe to merge :+1:

**Comment by Ma27** [comment]
and btw, thanks a lot for your effort!

**Pull request merged**